internal/states/resource.go

```diff
@@ -38,6 +38,42 @@ func (rs *Resource) Instance(key addrs.InstanceKey) *ResourceInstance {
 	return rs.Instances[key]
 }
 
+// InstanceProvider returns the calculated provider for the instance with the given key. It also returns an indication
+// whether the provider was set on the resource level or the instance level
+func (rs *Resource) InstanceProvider(key addrs.InstanceKey) (provider addrs.AbsProviderConfig, isSetOnInstanceLevel bool) {
+	var resourceProvider addrs.AbsProviderConfig
+	var instanceProvider addrs.AbsProviderConfig
+
+	resourceProvider = rs.ProviderConfig
+
+	// If the provider is set on the instance level, we can't get it from the resource
+	instance := rs.Instances[key]
+
+	if instance.Current != nil && instance.Current.InstanceProvider.Provider.Type != "" {
+		instanceProvider = instance.Current.InstanceProvider
+	} else {
+		// Take the provider from an arbitrary item from the deposed map
+		for _, deposedInstance := range instance.Deposed {
+			instanceProvider = deposedInstance.InstanceProvider
```

Comment:
What's going in here?

Reply:
I clarified this part. Let me know if [this](https://github.com/opentofu/opentofu/pull/1963/commits/3a615ba0091d655498db9ac0907a2cb4e9cb7feb) is better.
But overall, this is an assumption based on how the code behaved before this change. The provider was set on the resource level, so all the deposed objects had the same provider.

Reply:
If I understand correctly, instance pending for destruction could still reference different providers, isn't it?

The comment of the `Deposed` field:

> // Deposed, if len > 0, contains any remote objects that were previously
> 	// represented by the corresponding resource instance but have been
> 	// replaced and are pending destruction due to the create_before_destroy
> 	// lifecycle mode.